TypeScript support

lu4 opened the pull request:
This PR is a major rewrite for the library, it introduces TypeScript support for node-opencl. Suggested changes are not backward compatible, there are changes in package.json, and changes in both C++ backend interface and NodeJS frontend interface, also MR contains couple of bugfixes, also there are changes in ... ... ... everywhere. I think this PR requires broader attention and some time to sync with existing code base. Or at least something to start with...

Hope to hear out your comments!

Thank you in advance!

lmeyerov commented:
@lu4  - picking this back up 

Does this preserve API compatibility with non-ts users? if not, can it be done in a way that does so, such as alt calls?

lu4 commented:
Hi @lmeyerov, I'm having trouble with understanding what kind of compatibility are you asking about? Do you mean whether JavaScript developers would be able to use the provided api without the need of migrating to TypeScript? (If so than yes, they should have such ability, but this support might require couple of checks from my side and probably couple of updates. If this is the compatibility you are asking about then let me know)

> @lu4 - picking this back up
> 
> Does this preserve API compatibility with non-ts users? if not, can it be done in a way that does so, such as alt calls?

lmeyerov commented:
Yes, exactly:

-- can new node-opencl users use it without using typescript themselves?
-- were there no public api changes, so existing node-opencl users can keep using it with no changes?

lmeyerov commented:
If both are a yes, will help test + land soon after https://github.com/mikeseven/node-opencl/pull/69 . . If a no... bet we can figure it out :)

lu4 commented:
-- can new node-opencl users use it without using typescript themselves?
--> In theory yes, but appropriate checks / preparations should happen before the answer becomes a definitive yes

-- were there no public api changes, so existing node-opencl users can keep using it with no changes?
--> It is hard for me to remember at this point all of the changes, but off-top of my head I've fixed couple of bugs I've found myself which in theory may break backward compatibility. As for public interface I've tried to stick as close as possible to both the original `node-opencl` design and official Khronos spec, and so the places where I've found severe discrepancy between the two, I've tried to improve (if such improvement was possible), these changes may also break the backward compatibility. The changes I remember are:
- Int64 support (I made it little bit different, right now I don't remember how exactly I've implemented, but since then BigInt has became official part of ES standard which is better candidate than a pair of low / high numbers)
- I've made deprecated OpenCL functions available (it shouldn't affect existing users, there's no need to be so restrictive about deprecated APIs)

Maybe there's something else, I don't remember. But what I can say for sure is that most of these changes would most probably affect major version of node-opencl.

In any case I'm open to help with transition (full or parts) but it'll definitely take some time to happen...

lmeyerov commented:
Gotcha, thanks! We're seeing some fires on the new AWS G4s, which is ~4x cheaper than any of the others, so that shot up in priority. After we figure that out (1w?), will try this & report back!

lmeyerov commented:
Hi @lu4 , just got commit access & merged the non-TS branch, and happy to test & land this one when you think it's ready.

lmeyerov commented:
RE:Breaking changes, esp. for increased conformance & coverage , sounds worthwhile, we can just do a major version bump & call out in a changelog.

lu4 commented:
Let's align a plan, what changes do you need to get comfy?

lu4 commented:
@lmeyerov sounds cool, I'm experimenting with `embind`, `ffi`, `genapi`, `nbind` for other project these tools allow for automatic/semi-automatic bindings generation and look promising, but as I suspect we won't be throwing away existing code. So will signal you when ready, not promising anything, but hope to complete in couple of days

lmeyerov commented on src/svm.cpp:
```diff
@@ -133,11 +133,8 @@ NAN_METHOD(enqueueSVMFree) {
-
-  if(ARG_EXISTS(5) && info[5]->BooleanValue())
-      eventPtr = &event;
+  cl_event* eventPtr = &event;
```
unsafe here + after?

lu4 replied:
It should be safe, I was trying not to leave anything unsafe during implementation, off top of my head I didn't like the polymorphic return type, no developer likes surprises. Also `CHECK_ERR(err);` seems to throw an error upon non-null error, so the `if` clause is pointless its negative branch will never be executed so why bother creating a conditional return type?

lu4 replied:
It's safe  CHECK_ERR(err) throws exception so `info.GetReturnValue().Set(JS_INT(CL_SUCCESS));` will never be executed

lmeyerov commented on test/test.bindings.js:
```diff
@@ -8,23 +8,27 @@ describe("Equality", function() {
 
   it("should be falsy when trying to compare a NoCLObject with a scalar", function () {
     U.withContext(function (ctx) {
-      assert.equal(ctx.equals(5), false, "should be false");
```
looks like TODOs?

lu4 replied:
I don't remember what happened here, I suspect the test was failing and I didn't find no references of `equals` on the `ctx`, neither I remember what exactly is `ctx.equals()` I don't have clear picture what relation `ctx` has to `equals` why we'd ever want to compare things like that? Why not just `ctx1 === ctx2` or `ctx1 === 5` which would naturally be false.

lmeyerov commented on test/test.context.js:
```diff
@@ -41,74 +36,49 @@ describe("Context", function () {
-      var properties = [
-        cl.CONTEXT_PLATFORM, platform
-      ];
+    skip().vendor('nVidia').it("should create a context with the host processor type", function () {
```
We shouldn't be regressing on tests..

lu4 replied:
I agree

lmeyerov commented on test/test.device.js:
```diff
@@ -76,8 +75,8 @@ describe("Device", function() {
-      testString(device, "DEVICE_BUILT_IN_KERNELS");
-      testString(device, "DEVICE_SPIR_VERSIONS");
+      // testString(device, "DEVICE_BUILT_IN_KERNELS");
+      // testString(device, "DEVICE_SPIR_VERSIONS");
```
should not skip tests... better to guard tests by hw build?

lu4 replied:
I'll rethink this

lmeyerov commented on test/test.event.js:
```diff
@@ -27,7 +27,7 @@ describe("Event", function() {
   describe("#getEventInfo",function(){
     function testNumber(info,name,expected) {
-      it("should return the good value for " + name, function () {
+      skip().vendor("nVidia").it("should return the good value for " + name, function () {
```
more skipped tests

lu4 replied:
I see more need of incremental migration to TypeScript support.

lmeyerov commented on test/test.kernel.js:
```diff
@@ -138,11 +140,11 @@ describe("Kernel", function () {
 
-          if (cl.VERSION_1_2) {
-            U.bind(cl.setKernelArg, k, 0, [5, 10, 15])
+          if (cl.v12) {
```
more v12 refactors, not sure why

lu4 replied:
I can return back the naming but except for naming itself there are changes related to contents of namespaces, `node-opencl` does not provide deprecated APIs on javascript side, it literally excluded during compile time, I think this is wrong thing to do, if API is there it should be available. As for reasons to change naming is that I find 'v12' more convenient to write, no need to shout `VERSION (press shift) (press minus) 1 (press shift) (press minus) 2`

lmeyerov commented on test/test.kernel.js:
```diff
@@ -155,10 +157,10 @@ describe("Kernel", function () {
+            assert(cl.setKernelArg(k, 2, null, 5) == cl.SUCCESS);
           }
-          assert(cl.setKernelArg(k, 2, 5, "uint") == cl.SUCCESS);
+          assert(cl.setKernelArg(k, 2, "uint", 5) == cl.SUCCESS);
```
looks like kernel arg order changed, call out?

lu4 replied:
good point

lmeyerov commented on test/test.memobj.js:
```diff
@@ -402,47 +404,47 @@ describe("MemObj", function() {
     });
 
-    it("should get supported image formats", function () {
+    skip().vendor('nVidia').it("should get supported image formats", function () {
```
Why skip?

lu4 replied:
don't have an answer now

lmeyerov commented:
Wow, went through all core files and most of tests, this is great!

AFAICT, this:

-- Adds TS 🚀🚀🚀
-- More nan upgrades 💪💪💪
-- Regresses on callback support?
-- Disables a bunch of tests?  😔
-- Some conformance fixes: some arg orders, ... 💪💪
-- A few unnecessary refactors: `CL_1_2` => `v12`, getInfo signature change, ...
-- Some file moves, some of which aren't version-tracked
-- README/package.json changes from while this was a fork

My thinking to get landed:
-- Need to document the breaking API-level changes.  IMO, fine to do a least-effort solution like a migration section at the bottom of the main readme. 
-- May be smart to back out or split into a diff PR non-TS breaking changes ones such as the .info -- low value, mostly incurred work on others (e.g., we use that for some user-level scheduling, so annoying to port due to low value)
-- What's going on w/ the event args? Looks like they got cut some reason?
-- OK to lose version history on test/examples
-- Tests should be reenabled
-- I'm not a great reviewer of some of the C++ stuff, which compounds pain of a massive PR, but this is hard piecemeal so I get it :)

So maybe the flow is:
-- Think about ^^^^
-- Address what is reasonable
-- When ready, I can help basic AWS nvidia linux testing + osx intel
-- Once we're happy with that, I can smoke test on our full production workloads
-- ... and help with npm publishing (will figure out that flow w/ current branch)
-- If the regressions get too hard to land immediately, we can start by doing a TS branch

Sounds reasonable / ideas?

lmeyerov commented:
> @lmeyerov sounds cool, I'm experimenting with `embind`, `ffi`, `genapi`, `nbind` 

Cool! Yeah, I'm cool with most things that eliminate code and improve maintainability. Some of those look like quite risky dependencies that'll bite us in 6mo so rather not take on the likely coming tech debt, but beyond that, am supportive.

lu4 commented:
Ok, I'm back, will review and respond to your comments, will create a checklist and gradually will fix them...

lu4 commented on .vscode/launch.json:
```diff
@@ -0,0 +1,55 @@
+{
+    // Use IntelliSense to learn about possible attributes.
```
This is default Visual Studio Code `launch.json` config file, I think it is reasonable to remove them

lu4 commented on README.md:
```diff
@@ -1,3 +1,43 @@
+
+I've taken the initiative to renovate the existing code for my own purposes and now I'm sharing the result in hope that it will be useful for anyone else who is lacking typescript support with `node-opencl`. 
+
+This project contains full typings coverage of OpenCL 1.0-2.0 (and OpenCL 2.1, 2.2 ready to be implemented) specification (based on Khronos specification) with node-opencl specifics taken into account. The package also contains couple of minor bugfixes suggested by community PRs in original repo and couple of mine aesthetic interface changes / additions and changes related to management of deprec ...
```
I think we can stick to original readme, or have new one created, no need for message above

lu4 commented on package.json:
```diff
@@ -1,11 +1,13 @@
 {
-  "name": "node-opencl",
-  "version": "0.4.5",
+  "name": "node-opencl-ts",
```
Will rename this to original package name

lu4 commented on package.json:
```diff
@@ -38,21 +37,30 @@
-    "memwatch-next": "^0.2.10",
-    "nan": "^2.1.0",
-    "pngjs2": "*"
+    "nan": "^2.13.2"
```
What's your specific suggestion in relation to this change?

lu4 commented on package.json:
```diff
@@ -38,21 +37,30 @@
+    "tslint-config-prettier": "^1.17.0"
+  },
+  "peerDependencies": {
+    "node-memwatch": "1.0.1"
```
ok, no problem will remove it

lu4 commented on src/addon.cpp:
```diff
@@ -437,7 +504,7 @@ NAN_MODULE_INIT(init)
   JS_CL_CONSTANT(RGx);
   JS_CL_CONSTANT(RGBx);
-#ifdef  CL_VERSION_1_2
+#ifdef  v12
```
There are couple of things I wanted to address in relation to macros, first that macros shouldn't take responsibility and turn off deprecated OpenCL features just because they are deprecated, if there's a feature and works it should be available through the binding

lu4 commented on binding.gyp:
```diff
@@ -44,15 +44,23 @@
-            'OPENCL_SDK' : '<!(echo $AMDAPPSDKROOT)',
-            'OPENCL_SDK_INCLUDE' : '<(OPENCL_SDK)/include',
-            'OPENCL_SDK_LIB' : '<(OPENCL_SDK)/lib/x86_64',
+            'AMD_OPENCL_SDK' : '<!(echo $AMDAPPSDKROOT)',
```
Yes, I've just noticed, it's not just renaming

lu4 commented on src/commandqueue.cpp:
```diff
@@ -41,15 +34,19 @@ namespace opencl {
 NAN_METHOD(CreateCommandQueue) {
   Nan::HandleScope scope;
-  REQ_ARGS(3);
+  REQ_ARGS(2);
```
No, only 2 args are required, the third one is optional, here look at usage: https://github.com/mikeseven/node-opencl/pull/70/files/9dc971eafd3647a50aebbcfe85a346e7fbaa13f5#diff-df22d23427ce91c3cbe709d1e45371ecR48 and  here are specs https://www.khronos.org/registry/OpenCL/sdk/1.2/docs/man/xhtml/clCreateCommandQueue.html there's no need to specify null every time

lu4 commented on src/commandqueue.cpp:
```diff
@@ -1165,7 +1158,7 @@ NAN_METHOD(EnqueueMigrateMemObjects) {
 NAN_METHOD(EnqueueNDRangeKernel) {
   Nan::HandleScope scope;
-  REQ_ARGS(6);
+  REQ_ARGS(5);
```
No, there are only 5 required arguments the other ones are optional here look at usage: https://github.com/mikeseven/node-opencl/pull/70/files/9dc971eafd3647a50aebbcfe85a346e7fbaa13f5#diff-ef185681bfe4e868502e537b34b6f97fR120

lu4 replied:
no need to pass array of events when you don't want to, it's ok if parameter not specified, it'll be NULL on the C++ side anyway

lu4 commented on src/context.cpp:
```diff
@@ -78,7 +79,7 @@ NAN_METHOD(CreateContext) {
 NAN_METHOD(CreateContextFromType) {
   Nan::EscapableHandleScope scope;
-  REQ_ARGS(4)
+  REQ_ARGS(2) // Callbacks are not currently supported, neither user_data argument
```
Looking to how many changes are there, I'm thinking about incremental changes to this repository, the changes need a logical break-down. Also I think it's reasonable in place of callbacks to provide native promise implementations and have async / await for these methods on the JavaScript side, such change should be carefully re-thinked since queuing callback will cause a work to be executed in the background thread which may be an unexpected behavior

lu4 commented on src/device.cpp:
```diff
@@ -195,28 +204,21 @@ NAN_METHOD(GetDeviceInfo) {
-    arr->Set(0, JS_INT((uint32_t) (param_value>>32))); // hi
-    arr->Set(1, JS_INT((uint32_t) (param_value & 0xffffffff))); // lo
-    info.GetReturnValue().Set(arr);
+    info.GetReturnValue().Set(JS_INT((uint32_t) (param_value >> 10)));
```
There are just couple of places where Int64's are used most of them seem insignificant, the reason I've changed this is that it's inconvenient to work with. I don't know whether it should be reverted or changelog'ed? What'r your thoughts?

lu4 commented:
Ok, I've reviewed all your comments, I see that it might be too hard to try to move the new thing on top of old thing, there are too many layers of changes. Let's do it this way, I see that it's possible to introduce TypeScript bindings into existing project without touching the C++ side. I'll do this in a separate PR, and as of that we can start moving towards C++ changes (if necessary) 

I'm saying "if necessary" because I'd rewrite the whole thing from scratch, I don't like most of the things I see, but unfortunately I don't have time for such adventure. So let's go forward with this step by step... What are your thoughts on this?

lmeyerov commented:
Yes, splitting it up would make this a lot more approachable!

Maybe something like:

1. Prioritize an initial TS PR as a minor semvar bump:
-- Make the TS match the existing signatures. If any breaking changes, make an explicit and well-motivated migration list
-- I can help turn the list into an upgrade guide + help w/  device & integration testing
-- => Most people can start using with high confidence and upgrade with little pain

2a. C++ changes  (nan, ...) as a diff PR, ideally also as a minor semvar that conforms to (1)'s TS
=> same benefits as (1), and easier to do the security etc. review

2b. New abstractions (eg., async/await) as a major semvar, and maybe have live as an @next for awhile => "node-opencl 2.0"
